src/components/Stories/Stories.tsx

```diff
@@ -69,94 +51,61 @@ export function Stories({
     );
 
     const handleGotoPrevious = React.useCallback(() => {
-        if (currentStoryIndex > 0) {
+        setStoryIndex((currentStoryIndex) => {
+            if (currentStoryIndex <= 0) {
+                return 0;
+            }
+
             const newIndex = currentStoryIndex - 1;
-            setCurrentStoryIndex(newIndex);
             onPreviousClick?.(newIndex);
-        }
-    }, [currentStoryIndex, onPreviousClick]);
+            return newIndex;
+        });
+    }, [onPreviousClick]);
 
     const handleGotoNext = React.useCallback(() => {
-        if (currentStoryIndex < items.length - 1) {
+        setStoryIndex((currentStoryIndex) => {
+            if (currentStoryIndex >= items.length - 1) {
+                return items.length - 1;
+            }
+
             const newIndex = currentStoryIndex + 1;
-            setCurrentStoryIndex(newIndex);
             onNextClick?.(newIndex);
-        }
-    }, [currentStoryIndex, items, onNextClick]);
+            return newIndex;
+        });
+    }, [items, onNextClick]);
+
+    if (items.length === 0) {
+        return null;
+    }
+
+    // case when items has changed and index has ceased to be valid
+    if (items[storyIndex] === undefined) {
+        const correctIndex = items[initialStoryIndex] === undefined ? 0 : initialStoryIndex;
+        setStoryIndex(correctIndex);
+
+        return null;
```

Comment:
Why does it return null here?

Reply:
cause current `storyIndex` is invalid and we expect that `setStoryIndex(correctIndex);` will call render with valid index after `return null`.

Reply:
it seems like recursion, but not really
cause of [this lines](https://github.com/gravity-ui/uikit/blob/d3e88f692c195db489ea79debbbf53c568c621ca/src/components/Stories/Stories.tsx#L78)